File: aspenmq/src/main/java/net/aspenmq/transport/protocol/ConnectAck.java
package net.aspenmq.transport.protocol;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufInputStream;
import io.netty.buffer.ByteBufOutputStream;
import io.netty.buffer.Unpooled;

import java.io.IOException;

import net.aspenmq.transport.frame.FrameHeader;
import net.aspenmq.transport.frame.FrameHeader.QoS;
import net.aspenmq.transport.frame.MessageType;

public class ConnectAck  implements ProtocolMessage {
    public static final int CONNECTION_ACCEPTED = 0x00;
    public static final int ERROR_PROTOCOL_VERSION = 0x01;
    public static final int ERROR_IDENTIFIER_REJECTED = 0x02;
    public static final int ERROR_SERVER_UNAVAILABLE = 0x03;
    public static final int ERROR_BAD_CREDENTIALS = 0x04;
    public static final int ERROR_UNAUTHORIZED = 0x05;

    private int returnCode;
    int getReturnCode() {
        return returnCode;
    }
    void setReturnCode(int returnCode) {
        this.returnCode = returnCode;
    }

    @Override
    public ByteBuf encode() throws IOException {
        FrameHeader frameHeader = new FrameHeader(false,
                QoS.QOS_RESERVED,
                false,
                MessageType.CONNACK,
                2);
        byte[] headerBuf = new byte[FrameHeader.FIXED_HEADER_MAX_LENGTH];
        int headerLength = frameHeader.marshalHeader(headerBuf);

        ByteBuf buf = Unpooled.buffer(4);
        buf.writeBytes(headerBuf, 0, headerLength);
        ByteBufOutputStream bos = new ByteBufOutputStream(buf);
        bos.writeByte(0);
        bos.writeByte(returnCode);
        bos.flush();
        return buf;
    }

    public static ConnectAck decode(ByteBuf buf) throws IOException {
        ConnectAck connack = new ConnectAck();
        ByteBufInputStream bis = new ByteBufInputStream(buf);
        bis.readByte();
        connack.returnCode = bis.readByte();
        return connack;
    }
}

File: aspenmq/src/main/scala/net/aspenmq/transport/protocol/ConnectAck.scala
package net.aspenmq.transport.protocol

import io.netty.buffer.{ByteBuf, ByteBufInputStream, ByteBufOutputStream, Unpooled}
import net.aspenmq.transport.frame.{SFrameHeader, SMessageType, SQoS}

object ConnectAck {
  val CONNECTION_ACCEPTED = 0x00
  val ERROR_PROTOCOL_VERSION = 0x01
  val ERROR_IDENTIFIER_REJECTED = 0x02
  val ERROR_SERVER_UNAVAILABLE = 0x03
  val ERROR_BAD_CREDENTIALS = 0x04
  val ERROR_UNAUTHORIZED = 0x05

  def decode(buf: ByteBuf): ConnectAck = {
    val bis = new ByteBufInputStream(buf)
    try {
      bis.readByte()
      val returnCode = bis.readByte()
      new ConnectAck(returnCode)
    } finally {
      bis.close()
    }
  }
}

class ConnectAck(val returnCode: Int) extends ProtocolMessage with VariableHeaderEncoder {
  def encode(): ByteBuf = {
    val buf = Unpooled.buffer(2)
    val bos = new ByteBufOutputStream(buf)
    try {
      bos.writeByte(0)
      bos.writeByte(returnCode)
      bos.flush()
    } finally {
      bos.close()
    }
    val frameHeader = new SFrameHeader(false, SQoS.QOS_RESERVED, false, SMessageType.CONNACK, 2)
    super.encode(frameHeader, buf)
  }
}
